job-manager: support job hold and expedite

chu11 opened the pull request:
Per #2895 and #3324, support the idea of a job hold and job expedite.  A hold is urgency=0, in which a job will not be sent to the scheduler (note that #3334 still exists, so if the job has already been sent to the scheduler in unlimited mode, no way to change it yet).  Expedite is urgency=31, which automatically leads to the job priority being set to the max (2^32-1).

Supporting this in `flux job urgency`, the special args of `hold`, `expedite`, and `default` are now supported, making it easier for a
user to set a hold/expediate/default job priority on a job without knowing what values those map too.

Per discussion in #2895 it would be nice to inform a caller what the prior urgency was after changing it.  So I modified the `job-manager.urgency` RPC to return the old urgency to the caller.

I support a new `libjob` function `flux_job_set_urgency_get_prior_urgency()` to parse the RPC response.
I'm not a fan of this function name, and had trouble coming up with a good name.  Suggestions welcome.  Could just shorten to `flux_job_set_urgency_get()`, which is what I also considered.  Or need to shorten `flux_job_set_urgency()` to flux_job_urgency()` too.

I then support a `--verbose` option in `flux job urgency` to output a message to stderr on what the prior urgency was.  I had trouble with the option name here too.  Could have been `--output-prior` or something like that (in which I could perhaps output to stdout instead of stderr), but just settled on verbose in case we add more output in the future.  Suggestions welcome as well.

grondo commented:
> I support a new libjob function `flux_job_set_urgency_get_prior_urgency()` to parse the RPC response.

Hrm, I agree we need a better name! Why not drop the `prior_urgency` suffix and just call it `flux_job_set_urgency_get()`, it is not like there's going to be any other calls to process a set_urgency response...

Not sure if there's a use case for this, but it seems like there should be a way to query the current urgency without modifying it. I suppose this will fall out simply by supporting `+/-N`, if you made the default payload `{"op": "increment", "value": 0}`.

grondo commented on t/job-manager/sched-helper.sh:
```diff
@@ -43,7 +43,7 @@ jmgr_check_state() {
         local id=$(flux job id $1)
         local key=$2
-        local note="$(${JMGR_JOB_LIST} | grep ${id} | cut -f 6- | jq ."${key}")"
+        local note="$(${JMGR_JOB_LIST} | grep ${id} | cut -f 7- | jq ."${key}")"
```
This kind of thing in tests makes me cringe. Could we update the test job listing helper to optionally restrict output by key directly and get rid of the brittle `cut -f N-` used here? Anything to help a future test writer...

chu11 replied:
yeah, i concur.  Let me see if it'd be a huge effort to redo that tool and all the tests associated with it.

grondo commented on src/cmd/flux-job.c:
```diff
@@ -127,6 +127,13 @@ static struct optparse_option list_inactive_opts[] =  {
 
+static struct optparse_option urgency_opts[] =  {
+    { .name = "verbose", .key = 'v', .has_arg = 0,
+      .usage = "Increase verbosity",
```
```suggestion
      .usage = "Output prior urgency value on success",
```

chu11 commented:
> Hrm, I agree we need a better name! Why not drop the prior_urgency suffix and just call it flux_job_set_urgency_get(), it is not like there's going to be any other calls to process a set_urgency response...

Good point, and that actually falls through into ...

> Not sure if there's a use case for this, but it seems like there should be a way to query the current urgency without modifying it. I suppose this will fall out simply by supporting +/-N, if you made the default payload {"op": "increment", "value": 0}.

After thinking about #3427 for a few minutes this morning, I decided to just go ahead and add the +/-N support since its sort of just falls into line with this PR.  We can return both the old & new value in verbose output.

garlick commented:
I commented over in #3334, but it seems like it would be fine to cancel an outstanding alloc request for job that is being held.  If that's the way to go, then should it be done in this PR?  (o/w hold is fairly non-functional for our production scheduler that uses mode=unlimited)

chu11 commented:
> I commented over in #3334, but it seems like it would be fine to cancel an outstanding alloc request for job that is being held. If that's the way to go, then should it be done in this PR? (o/w hold is fairly non-functional for our production scheduler that uses mode=unlimited)

You're speaking about the mode=unlimited case?  Yeah, I didn't really think about that case, figuring it'd be handled in #3334.  But I suppose it can be added here.

garlick commented:
Do we think exposing "urgency" in the API is going to be useful?  I know we have `flux_job_set_urgency()` now, but before tweaking it for getting the previous value, or for making relative vs absolute changes, maybe we should ask whether `flux job urgency` is likely to be the only user, and leave it out if so.

I don't know if this would be a thing, but I could imagine a workflow that submits held jobs and releases them once some workflow condition is met.  Although maybe that would be a job for dependencies once we have those...

If we do want API support, maybe we ought to go wtih something generic like
```c
flux_future_t *flux_job_modify_pack  (flux_t *h, flux_jobid_t id, const char *name, const char *fmt, ...);
int flux_job_modify_get_unpack (flux_future_t *f, const char *fmt, ...);
```
and then support (for example):
```c
f = flux_job_modify_pack (h, id, "urgency", "{s:b s:i}", "relative", 1, "value", -1);
flux_job_modify_get_unpack (f, "{s:s}", "value", &last_value);
```
I don't know if that's much better than having people make the RPCs directly though, since we'd need to carefully document the name-specific arguments.

Just throwing that out there as an idea.

grondo commented:
I like the idea of `flux_job_modify()` in general. The amount of enums and API interfaces in `job.h` has kind of got me down (considering any future changes there would break ABI and have to be very carefully considered). I like the flexibility of a `modify` api with named args...

In fact, this might be very useful once there is a plugin interface in the job-manager, perhaps the "modify" event can be passed into the plugin to modify plugin specific data about jobs. (useful at least for testing, perhaps also allowing more powerful plugins to be developed in the future)

garlick commented:
oK, given that, would it make sense for this  PR to drop `flux_job_set_urgency()` and make direct RPCs for now?
Then we could open another issue on `flux_job_update()` and that could be another PR?

grondo commented:
Sounds good to me

chu11 commented:
> I like the idea of flux_job_modify() in general. 

I like this idea as well, and it flows into issue #3391 (which perhaps we can update to include `flux_job_modify()`)

chu11 commented:
Rebased and re-pushed.  Removed the cleanup patches that were merged in via #3436 and updated some tests to be cleaner as a result.

Based on discussion above.

- I no longer support a `flux_job_set_urgency_get_prior_urgency()`.    Instead I parse the response from `flux_job_set_urgency()` directly.

- I didn't remove `flux_job_set_urgency()`, feeling that is for another PR.

- The big addition is commit facdc08558ef4ab50f15ce12fad99723497b7592  ("job-manager: pack/unpack priority with json_int_t").  After  merging #3436, I discovered that the priority value -1 was sent via the `job-manager` list service but not with the `job-info` list service when priority=4294967295.

  When packing with jansson and `{s:i}`, this uses ints (4294967295 = MAX_UINT = -1 signed int).  As long as we unpack into unsigned int on the other end, this wasn't really a problem.  But when we converted the `list-jobs` testing tool to output json and not be parsed by C, the problem was exposed.  The reason this didn't occur in `job-info`, is b/c it constructed responses using `json_integer()`, which takes `json_int_t` (a long long).

  I later realized that when priority=4294967295, -1 was also stored for the priority in the KVS.

  There were several ways to deal with this.  I decided to solve this by simply packing/unpacking priorities with `{s:I}` (long long) instead of `{s:i}` (int).

chu11 commented:
re-pushed a fix, forgot a HAVE_JQ prereq

chu11 commented:
Doh, i just realized, i never addressed the cancellation of jobs in unlimited mode.

chu11 commented:
> I commented over in #3334, but it seems like it would be fine to cancel an outstanding alloc request for job that is being held. If that's the way to go, then should it be done in this PR? (o/w hold is fairly non-functional for our production scheduler that uses mode=unlimited)

After attempting to add this into this PR, I think its best to leave this for #3334.  it would involve more changes than I think would be worth (and without doing #3334).  I had forgotten about this chunk of code:

```
    /* RFC 27 does not yet handle urgency changes after alloc request
     * has been sent to the scheduler.  Also, alloc_queue_reorder() will
     * segfault if job->handle is NULL, which is the case if the job is
     * no longer in alloc->queue.
     */
    if (job->alloc_pending) {
        errstr = "job has made an alloc request to scheduler, "
                 "urgency cannot be changed";
        errno = EINVAL;
        goto error;
    }
```

So I think the current code is good for review (although at this point, it can be after the break :-).

chu11 commented:
> >  I commented over in #3334, but it seems like it would be fine to cancel an outstanding alloc request for job that is being held. If that's the way to go, then should it be done in this PR? (o/w hold is fairly non-functional for our production scheduler that uses mode=unlimited)
> 
> After attempting to add this into this PR, I think its best to leave this for #3334. it would involve more changes than I think would be worth (without basically starting to do #3334). I had forgotten about this chunk of code:

I take it back, I was able to think of a way to add it that wasn't too intrusive.  So holds in unlimited are now handled and some tests were added too.

garlick commented:
Just having a fresh look at this one.  If I followed the discussion, it sounds like all the review comments were addressed.  If there are specific areas I should focus on please let me know @chu11.

chu11 commented:
@garlick   I'm not sure anyone commented on this commit, as it could be viewed as a controversial solution.

https://github.com/flux-framework/flux-core/pull/3428/commits/36770a9e65f66801c6fa194da48f6485a218a6a8

grondo commented:
> @garlick I'm not sure anyone commented on this commit, as it could be viewed as a controversial solution.

Ah, good call. Storing `unsigned int` in JSON with negative values is a big bummer, especially when looking at eventlogs (it is already hard enough to tell what the priority means with large values before throwing negative values into the mix)

Therefore, I prefer your solution of storing `priority` with `"I"` as `json_int_t`.

The other thing I was wondering is why not just make `job->priority` `int64_t`. It would make the JSON encoding brainless, and avoid unpacking into temporary variables. I guess the drawback is that we waste 32bits in the job structure (but then again, this would allow us to specify an "unset" value for `job->priority`). But `urgency` only requires 5 bits and we're using 32 there, so there's already a precedent.. :wink:

chu11 commented:
> The other thing I was wondering is why not just make job->priority int64_t. It would make the JSON encoding brainless, and avoid unpacking into temporary variables. 

I just defaulted to storing in unsigned int b/c we define priority to be 0 to UINT_MAX in the RFC.  I could definitely store it as int64_t, but sort of like the range to match the priority range.

Hmmm, should I actually use `uint32_t` everywhere to be more clear?

> I guess the drawback is that we waste 32bits in the job structure (but then again, this would allow us to specify an "unset" value for job->priority). But urgency only requires 5 bits and we're using 32 there, so there's already a precedent.. wink

Don't quite follow the "unset" value comment.  i.e. we have some extra bits so we can utilize them as flags?

Should I dare to pose this question.  What's stopping us from re-defining priority to be 0 to LONG_MAX in the RFC??  Or if we want to save some bits for some special case, define as INT64 and say range is 0-2^48, saving 16 bits for the future just in case?

grondo commented:
> Don't quite follow the "unset" value comment. i.e. we have some extra bits so we can utilize them as flags?

Sorry that was unclear. Currently valid values for priority take up the whole range of the type. If we instead used int64_t, but still defined valid priorities in the same range, then we could set `priority = -1` to define and invalid or unset (not yet set) priority (e.g. as would be the case before the PRIORITY state)

> I could definitely store it as int64_t, but sort of like the range to match the priority range.

That was my point about a precedent, the urgency range is already only 0-32, which *doesn't* take up the entire range of the type. So doing so for priority can be seen as pretty arbitrary.

garlick commented:
Bah, yeah, let's just change the RFC.  Maybe it will be handy to have the extra range available.

chu11 commented:
> Bah, yeah, let's just change the RFC. Maybe it will be handy to have the extra range available.

keep the same range but make int64_t for future consideration?  Or make range 0-LONG_MAX?

garlick commented:
I meant extend the range to 0 to LONGLONGMAX-1.

grondo commented:
> Or make range 0-LONG_MAX?

FWIW, I was only arguing for storing priority _internally_ as 64bit signed to avoid problems with JSON encoding (and to allow `-1` to be an invalid value). 

Extending the range seems fine and would allow more precision in how factors in a multi-factor calculation might contribute to the final priority (though IMO 0-2^32 is plenty of range for now)

chu11 commented:
> Edit: sorry, didn't see @grondo's comment, and didn't mean to sound dismissive. I don't have a strong opinion, although 4 extra bytes per job to avoid an encoding inconvenience seems slightly wasteful to me.

Ok.  Everyone seems ok with the unsigned int -> long long int mapping for JSON.  How about we keep code as is, and I'll add an RFC issue to just remind us of this discussion.  Perhaps some day we'll need those extra bits.

priorities being `unsigned int` predates this PR, so i'll write up an issue for that and we can consider that for another PR.

chu11 commented:
@mergifyio rebase

chu11 commented:
oh crud, i forgot about @garlick's "via -> vs" comment.  I'll fixup and re-pushed.

grondo commented:
> I don't have a strong opinion, although 4 extra bytes per job to avoid an encoding inconvenience seems slightly wasteful to me.

Well it isn't just an encoding inconvenience. There is no way to store an invalid priority when the valid values take up the entire type. I guess I'm just repeating myself, but it doesn't matter that much to me. I've just personally experienced the "inconvenience" of "I" vs "i" in jansson way too many times (random corruption), and given that plugin authors may also have to encode/decode the priority, I thought it wise to make it as fool proof as possible. (also didn't worry about the extra 4 bytes per job since we're already storing jobspec for every job anyway, it seemed like a pittance)

This is probably not a big deal, we'll likely know soon after if people get it wrong.

garlick commented:
I'm for making it 64 bits, and changing the range to 63 bits (allowing a negative value to indicate "invalid" if needed).

I just thought if we're going to use the space, may as well make it available.

grondo commented:
Should we do the same for urgency? :stuck_out_tongue_winking_eye:

chu11 commented:
rebased & re-pushed, fixing the english in some comments that @garlick mentioned

garlick commented:
> Ok. Everyone seems ok with the unsigned int -> long long int mapping for JSON. How about we keep code as is, and I'll add an RFC issue to just remind us of this discussion. Perhaps some day we'll need those extra bits.

My vote would be change it to `int64_t` here and keep the RFC range, to avoid extra churn in the history, and probably an extra PR/release notes entry.

garlick commented on src/modules/job-manager/alloc.c:
```diff
@@ -63,7 +63,7 @@ static void requeue_pending (struct alloc *alloc, struct job *job)
     assert (job->handle == NULL);
     if (!(job->handle = zlistx_insert (alloc->queue, job, fwd)))
-        flux_log_error (ctx->h, "%s: queue_insert", __FUNCTION__);
+        flux_log_error (ctx->h, "%s: zlistx_insert", __FUNCTION__);
```
If we're fixing error messages here
* `xlistx_insert()` doesn't claim to set errno, so `flux_log_error()` is inappropriate
* how about something more descriptive like "failed to enqueue job for scheduling" and drop  `__FUNCTION__`

garlick commented on src/common/libjob/job.h:
```diff
@@ -31,6 +31,11 @@ enum job_urgency {
     FLUX_JOB_URGENCY_MAX = 31,
 };
 
+enum job_urgency_cmds {
```
can these just be added to the `job_urgency` enum?

garlick commented on src/modules/job-manager/alloc.c:
```diff
@@ -508,7 +510,8 @@ static void check_cb (flux_reactor_t *r, flux_watcher_t *w,
         return;
     if (alloc->mode == SCHED_SINGLE && alloc->alloc_pending_count > 0)
         return;
-    if ((job = zlistx_first (alloc->queue))) {
+    if ((job = zlistx_first (alloc->queue))
+        && job->urgency != FLUX_JOB_URGENCY_HOLD) {
```
Held job will prevent jobs after it in the queue from being sent to scheduler?

garlick commented on src/modules/job-manager/alloc.c:
```diff
@@ -483,12 +483,14 @@ static void prep_cb (flux_reactor_t *r, flux_watcher_t *w,
         return;
     if (alloc->mode == SCHED_SINGLE && alloc->alloc_pending_count > 0)
         return;
-    if (zlistx_first (alloc->queue))
+    if ((job = zlistx_first (alloc->queue))
+        && job->urgency != FLUX_JOB_URGENCY_HOLD)
         flux_watcher_start (alloc->idle);
```
If a held job ends up at the top of the queue, but there are other jobs underneath, won't this fail to wake up the reactor?

chu11 replied:
Unless I have a bug somewhere, I don't think so?  B/c the queue is sorted, if the job at the head of the queue has urgency == 0, then presumably all jobs after it are also urgency == 0.

garlick replied:
That makes sense, thanks, I didn't see that at first.  It might be good to add a comment to top of `prep_cb()` and `check_cb()`, something like
```c
/* The queue is sorted from highest to lowest priority, so if the
 * first job has urgency=HOLD (priority=MIN), all other jobs must have
 * the same priority, and no alloc requests can be sent.
 */
```

garlick replied:
For the record, I did manually verify that submitting a held job doesn't block things up, though your explanation makes it obvious that the code is correct.

garlick commented on src/modules/job-manager/urgency.c:
```diff
@@ -120,7 +123,8 @@ void urgency_handle_request (flux_t *h,
         if (alloc_queue_recalc_pending (ctx->alloc) < 0)
-            flux_log_error (h, "%s: alloc_queue_recalc_pending", __FUNCTION__);
+            flux_log_error (h, "%s: alloc_queue_recalc_pending",
+                            __FUNCTION__);
```
There will already be an error logged (two actually) if `alloc_queue_recalc_pending()` fails, and I thnk the only way it can fail is if a fire and forget RPC fails (like from ENOMEM).  Let's skip logging an error here too.

garlick commented on src/modules/job-manager/urgency.c:
```diff
@@ -125,10 +127,21 @@ void urgency_handle_request (flux_t *h,
-                            __FUNCTION__);
+        if (job->alloc_queued) {
+            alloc_queue_reorder (ctx->alloc, job);
+            if (alloc_queue_recalc_pending (ctx->alloc) < 0) {
+                flux_log_error (h, "%s: alloc_queue_recalc_pending",
+                                __FUNCTION__);
+                goto error;
+            }
```
Error message is redundant.

garlick commented on src/modules/job-manager/urgency.c:
```diff
@@ -125,10 +127,21 @@ void urgency_handle_request (flux_t *h,
+                flux_log_error (ctx->h, "%s: alloc_cancel_alloc_request",
+                                __FUNCTION__);
+                goto error;
+            }
```
Error message is redundant.

garlick commented on src/cmd/flux-job.c:
```diff
@@ -337,7 +337,7 @@ static struct optparse_subcommand subcommands[] = {
     { "urgency",
       "[OPTIONS] id urgency",
-      "Set job urgency",
+      "Set job urgency (0-31, HOLD, EXPEDITE, DEFAULT)",
```
It's not clear from the usage message that the textual values are aliases for values in the 0-31 range.

Would it be more clear  to say
> Set job urgency (0-31), hold=0, default=16, expedite=31

garlick reviewed (changes requested):
Is it reasonable to go ahead and change `job->priority` to an `int64_t` in this PR?

Big question about blocking the alloc queue when a job is held.

Other stuff is fairly minor.

chu11 commented:
> Is it reasonable to go ahead and change job->priority to an int64_t in this PR?

I think its reasonable to.

I realized I forgot to use `json_int_t` with the alloc request and the sched hello RPCs, so gotta adjust there as well.

garlick commented:
> I realized I forgot to use json_int_t with the alloc request and the sched hello RPCs, so gotta adjust there as well.

Note that configure checks that `json_int_t` is 64 bits, so it's safe to use `int64_t`.

chu11 commented:
re-pushed with fixes per @garlick comments above.  Mostly addressed with fixups, but there's also one new commit (https://github.com/flux-framework/flux-core/pull/3428/commits/f05136a202d94c9e0a472070d98f8bc0ba2fd859) to deal with one of the fixes.

garlick reviewed (approved):
Looking good!

Feel free to squash the fixups and, I assume, the int64_t change with the earlier commits that use "I" with temp variables.

Edit: oh, and add that requested comment, if it makes sense.

chu11 commented:
just re-pushed, squashing everything and adding the comments @garlick mentioned above.  Identical to previous review except with these comments:

```
diff --git a/src/modules/job-manager/alloc.c b/src/modules/job-manager/alloc.c
index b9c24f1..1dc0817 100644
--- a/src/modules/job-manager/alloc.c
+++ b/src/modules/job-manager/alloc.c
@@ -489,6 +489,10 @@ static void prep_cb (flux_reactor_t *r, flux_watcher_t *w,
         return;
     if (alloc->mode == SCHED_SINGLE && alloc->alloc_pending_count > 0)
         return;
+   /* The queue is sorted from highest to lowest priority, so if the
+    * first job has urgency=HOLD (priority=MIN), all other jobs must have
+    * the same priority, and no alloc requests can be sent.
+    */
     if ((job = zlistx_first (alloc->queue))
         && job->urgency != FLUX_JOB_URGENCY_HOLD)
         flux_watcher_start (alloc->idle);
@@ -510,6 +514,10 @@ static void check_cb (flux_reactor_t *r, flux_watcher_t *w,
         return;
     if (alloc->mode == SCHED_SINGLE && alloc->alloc_pending_count > 0)
         return;
+   /* The queue is sorted from highest to lowest priority, so if the
+    * first job has urgency=HOLD (priority=MIN), all other jobs must have
+    * the same priority, and no alloc requests can be sent.
+    */
     if ((job = zlistx_first (alloc->queue))
         && job->urgency != FLUX_JOB_URGENCY_HOLD) {
         if (alloc_request (alloc, job) < 0) {

```

garlick commented:
Perfect, thanks @chu11. I'll set MWP.
